app/models/preorder_information.rb

```diff
@@ -8,6 +8,9 @@ class PreorderInformation < ApplicationRecord
 
   validates :status, presence: true
 
+  before_save :check_and_update_status_if_necessary
+  after_touch :refresh_status!
```

Comment:
Could we do `before_touch :set_status` or similar so that we needn't save twice?

Reply:
Unfortunately there is no `before_touch`. I don't think we'd be doing 2 saves here though as the `before_save` callback won't trigger the `after_touch`. The `after_touch` callback is only triggered by `touch`es and seems to be protected in a sensible way from getting into infinite loops.